drivers/serial/uart_gecko.c

```diff
@@ -0,0 +1,421 @@
+
+struct uart_gecko_config {
+	USART_TypeDef *base;
+	CMU_Clock_TypeDef clock;
+	uint32_t baud_rate;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(struct device *dev);
+#endif
+};
+
+struct uart_gecko_data {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_t callback;
+#endif
+};
+
+static int uart_gecko_poll_in(struct device *dev, unsigned char *c)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint32_t flags = USART_StatusGet(config->base);
+	int ret = -1;
+
+	if (flags & USART_STATUS_RXDATAV) {
+		*c = USART_Rx(config->base);
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static unsigned char uart_gecko_poll_out(struct device *dev, unsigned char c)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+
+	USART_Tx(config->base, c);
+
+	return c;
+}
+
+static int uart_gecko_err_check(struct device *dev)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint32_t flags = USART_IntGet(config->base);
+	int err = 0;
+
+	if (flags & USART_IF_RXOF) {
+		err |= UART_ERROR_OVERRUN;
+	}
+
+	if (flags & USART_IF_PERR) {
+		err |= UART_ERROR_PARITY;
+	}
+
+	if (flags & USART_IF_FERR) {
+		err |= UART_ERROR_FRAMING;
+	}
+
+	USART_IntClear(config->base, USART_IF_RXOF |
+		       USART_IF_PERR |
+		       USART_IF_FERR);
+
+	return err;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static int uart_gecko_fifo_fill(struct device *dev, const uint8_t *tx_data,
+			       int len)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint8_t num_tx = 0;
+
+	while ((len - num_tx > 0) &&
+	       (config->base->STATUS & USART_STATUS_TXBL)) {
+
+		config->base->TXDATA = (uint32_t)tx_data[num_tx++];
+	}
+
+	return num_tx;
+}
+
+static int uart_gecko_fifo_read(struct device *dev, uint8_t *rx_data,
+			       const int len)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint8_t num_rx = 0;
+
+	while ((len - num_rx > 0) &&
+	       (config->base->STATUS & USART_STATUS_RXDATAV)) {
+
+		rx_data[num_rx++] = (uint8_t)config->base->RXDATA;
+	}
+
+	return num_rx;
+}
+
+static void uart_gecko_irq_tx_enable(struct device *dev)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint32_t mask = USART_IEN_TXBL;
+
+	USART_IntEnable(config->base, mask);
+}
+
+static void uart_gecko_irq_tx_disable(struct device *dev)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint32_t mask = USART_IEN_TXBL;
+
+	USART_IntDisable(config->base, mask);
+}
+
+static int uart_gecko_irq_tx_complete(struct device *dev)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint32_t flags = USART_IntGet(config->base);
+
+	return (flags & USART_IF_TXBL) != 0;
+}
+
+static int uart_gecko_irq_tx_ready(struct device *dev)
+{
+	const struct uart_gecko_config *config = dev->config->config_info;
+	uint32_t mask = USART_IEN_TXBL;
+
+	return (config->base->IEN & mask)
+		&& uart_gecko_irq_tx_complete(dev);
```

Comment:
This isn't exactly correct. First of all "tx complete" and "tx ready" are 2 different events. "Tx ready" means "tx buffer is ready to accept a char". "Tx complete" means "all data in tx buffer was transferred". In general, it's not correct to express one in terms of another (there should be 2 different status bit in the hardware).

Another less serious issue is that there's no requirement for masking the condition in the software (so, `(config->base->IEN & mask)` is superfluous).

Note that it's a known issue that many existing drivers don't follow these guidelines, and fixing those is on TODO, per https://jira.zephyrproject.org/browse/ZEP-2016 .